File: sdks/rust/examples/basic.rs
//! Basic KV Store Example
//!
//! This example demonstrates basic key-value operations using the Synap SDK.
//!
//! Usage:
//!   cargo run --example basic

use synap_sdk::{SynapClient, SynapConfig};
use tracing::info;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt::init();

    // Create Synap client
    let config = SynapConfig::new("http://localhost:15500");
    let client = SynapClient::new(config)?;

    info!("🚀 Synap Rust SDK - Basic Example\n");

    // 1. SET a string value
    info!("1. Setting key 'greeting' = 'Hello, Synap!'");
    client.kv().set("greeting", "Hello, Synap!", None).await?;

    // 2. GET the value
    info!("2. Getting key 'greeting'");
    let value: Option<String> = client.kv().get("greeting").await?;
    info!("   Value: {:?}\n", value);

    // 3. SET with TTL (10 seconds)
    info!("3. Setting key 'session' with 10s TTL");
    client
        .kv()
        .set("session", "temporary-token", Some(10))
        .await?;

    // 4. Atomic operations
    info!("4. Increment counter");
    client.kv().set("counter", 0, None).await?;
    let val = client.kv().incr("counter").await?;
    info!("   Counter after INCR: {}", val);

    let val = client.kv().incr("counter").await?;
    info!("   Counter after INCR: {}", val);

    let val = client.kv().decr("counter").await?;
    info!("   Counter after DECR: {}\n", val);

    // 5. Check existence
    info!("5. Checking if key exists");
    let exists = client.kv().exists("greeting").await?;
    info!("   'greeting' exists: {}", exists);

    let exists = client.kv().exists("nonexistent").await?;
    info!("   'nonexistent' exists: {}\n", exists);

    // 6. Delete a key
    info!("6. Deleting key 'greeting'");
    let deleted = client.kv().delete("greeting").await?;
    info!("   Deleted: {}\n", deleted);

    // 7. Get statistics
    info!("7. Getting KV store statistics");
    let stats = client.kv().stats().await?;
    info!("   Total keys: {}", stats.total_keys);
    info!("   Total memory: {} bytes", stats.total_memory_bytes);
    info!("   Hit rate: {:.2}%", stats.hit_rate * 100.0);

    info!("\n✅ Example completed successfully!");

    Ok(())
}

File: sdks/php/examples/basic.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use Synap\SDK\SynapClient;
use Synap\SDK\SynapConfig;

// Create client
$config = SynapConfig::create('http://localhost:15500')
    ->withTimeout(30);

$client = new SynapClient($config);

echo "=== Synap PHP SDK - Basic Example ===\n\n";

// Key-Value operations
echo "1. Key-Value Store:\n";
$client->kv()->set('user:1', 'John Doe');
$value = $client->kv()->get('user:1');
echo "  - Set user:1 = John Doe\n";
echo "  - Get user:1 = {$value}\n";

// With TTL
$client->kv()->set('session:abc', 'token123', 3600);
echo "  - Set session with 3600s TTL\n";

// Atomic operations
$counter = $client->kv()->incr('visitors');
echo "  - Increment visitors = {$counter}\n";

echo "\n";

// Queue operations
echo "2. Message Queue:\n";
$client->queue()->createQueue('tasks');
$msgId = $client->queue()->publish('tasks', ['task' => 'process-video'], 9);
echo "  - Published message: {$msgId}\n";

$message = $client->queue()->consume('tasks', 'worker-1');
if ($message) {
    echo "  - Consumed message: {$message->id}\n";
    $client->queue()->ack('tasks', $message->id);
    echo "  - Acknowledged message\n";
}

echo "\n";

// Stream operations
echo "3. Event Stream:\n";
$client->stream()->createRoom('chat-room-1');
$offset = $client->stream()->publish('chat-room-1', 'message', [
    'user' => 'alice',
    'text' => 'Hello!',
]);
echo "  - Published event at offset: {$offset}\n";

$events = $client->stream()->consume('chat-room-1', 0, 10);
echo "  - Consumed " . count($events) . " events\n";

echo "\n";

// Pub/Sub operations
echo "4. Pub/Sub:\n";
$client->pubsub()->subscribeTopics('user-123', ['notifications.*']);
$delivered = $client->pubsub()->publish('notifications.email', [
    'to' => 'user@example.com',
    'subject' => 'Welcome',
]);
echo "  - Published to {$delivered} subscribers\n";

echo "\n✅ All operations completed successfully!\n";
